### DISCUSSION TITLE
USB PID/VID

### DISCUSSION THREAD

**Discussion by sstaub** [earlephilhower/arduino-pico#442] [Q&A]
Is there a way to change PID / VID of the USB Serial interface?

**Comment by john-4** [comment]
I'd like to know this as well, probably for the same reason as you sstaub, emulating the old Arduino boards compatible with console Lighthack.

Typically under Arduino compatibles (e.g. Teensy) this is done in a "boards.txt", and in Earle's arduino-pico core that file is seemingly located at _[...]\packages\rp2040\hardware\rp2040\<version>\boards.txt_ .

By editing that file, I've had success spoofing my board's PID to Lighthack's, but _not_ the VID. The usual fixes"<board>.build.vid=0x2341" "<board>.build.pid=0x0043" didn't do anything.   However, by changing "**<board>.build.usbpid=-DSERIALUSB_PID=_0x0043_**" from its original value, I was able to spoof my usb-serial PID.  There is no corresponding entry like "build.usbvid" though, therefore it looks like no current way to spoof the VID?

I'd really appreciate an official answer on this, as the RP2040 size, cost, and connectivity make it very appealing for Lighthack.  In the meantime I'll keep tinkering, it may be possible to spoof it using the regular boards.txt method when running the new official mbed-based rp2040 core.

**Reply by earlephilhower** [reply]
It's pretty straightforward to add this `#define` override.  Basically the same logic as is there for the PID.  Would you like to make a PR?

**Reply by john-4** [reply]
Wow, that was quick!

I'm afraid I'm not familiar enough with the buildchain to understand where to make that definition properly.  And I've never made a pull request on someone else's project before!

If you could let me know where/what to edit, I'll check that it works for my Eos console and then try a PR?

Thanks again!

**Reply by earlephilhower** [reply]
See https://github.com/earlephilhower/arduino-pico/blob/5be405308e31b864f0d8f213fdf266344ed45bce/cores/rp2040/RP2040USB.cpp#L48-L52

Do the same kind of ifdef/etc. for https://github.com/earlephilhower/arduino-pico/blob/5be405308e31b864f0d8f213fdf266344ed45bce/cores/rp2040/RP2040USB.cpp#L46

and you should be all set.

**Reply by earlephilhower** [reply]
And to make a PR you'd fork this repo from the GUI, clone it on your machine, do the edit and commit to a new branch.  The push the branch to your repo and go back to the web UI.  In your `code` (main) repo page you should see a button to let you make a PR on the recent branch push.

**Reply by minisbett** [reply]
Hey! Sorry for reviving this old discussion, but I looked up the current implementation of these defines and from my understanding I can override both the VID and PID by adding the build flags `-DUSBD_VID=0x0727 -DUSBD_PID=0x0727`, however this did not override anything. How can I override this properly?

**Reply by jcmartinez** [reply]
I'm also interested in this.

Do you mean that for each new PID / VID, we must submit a pull request?

Is there a way of overriding PID / VID, as needed?

I would like to change the following values on my board:

$ lsusb -v
  idVendor           0x2e8a 
  idProduct          0x000a 
  bcdDevice            1.00
  iManufacturer           2 Raspberry Pi
  iProduct                3 Pico

Someone at https://community.platformio.org/t/how-to-define-custom-usb-pid-vid-and-product-names-with-pico/33538 mentioned that they can override these values by creating a new board definition. My project is in VSCode using the Arduino Maker Workshop extension, and I haven't been able to create a new board definition.

Could you shed some light on this?

**Reply by earlephilhower** [reply]
Just define the custom USBD_xxx values in your build system.  In the IDE, you'd need a boards.local.txt or platform.local.txt.  In Platform.IO you can pass in the defines.
https://github.com/earlephilhower/arduino-pico/blob/dba8f173358d55b5f7ca2354f22e589a53275087/cores/rp2040/RP2040USB.cpp#L45-L51

**Reply by maxgerhardt** [reply]
We've also since then documented the PlatformIO settings part in https://arduino-pico.readthedocs.io/en/latest/platformio.html#usb-customization, for completenes.

**Comment by madias123** [comment]
Ok, I'm a little bit lost, struggling with 2 USB-Midi devices with your core and the classic Windows name mismatch.
So I set up a bards.local.txt file with
` vccgnd_yd_rp2040.build.extra_flags=-DUSBD_VID=0x1209 -DUSBD_PID=0x0002 -DUSBD_MANUFACTURER="m-diro" -DUSBD_PRODUCT="diro_audio" -DUSBD_SN="DIRO987654"`
but even after restart the IDE, deleting all old device entries in the device manager, it still not shown up in USBDeview:

- Serial Number stays empty
- VendorID: 239a
- ProductID: cafe

What I´m doing wrong?

**Reply by earlephilhower** [reply]
@maxgerhardt  and I are talking about updating the core USB.  Evidently you're not using the core, so you need to use the appropriate Adafruit defines.  Those should be documented in their docs.  If not look in `Adafruit_USBD_Device.cpp` to see the exact `#define`s you need to override.

**Comment by madias123** [comment]
...shame on me :/ 
Thank you - I completely forgot that I'm using the Adafruit TinyUSB stack. And VendorID: 239a IS Adafruit....
Coding time: 14h today - let's go to bed.

**Comment by earlephilhower** [comment]
For future reference, the core can now set any VID/etc. from your sketch.  See https://github.com/earlephilhower/arduino-pico/blob/5.2.0/libraries/rp2040/examples/USBConfigure/USBConfigure.ino
